File: 815-champagne-tower/champagne-tower.cpp
class Solution {
public:
    double champagneTower(int poured, int query_row, int query_glass) {
        vector<vector<double>> glass(102, vector<double>(101, 0.0));
        glass[0][0] = poured;
        for(int i = 0; i < query_row + 1; i++) {
            for(int j = 0; j < i + 1; j++) {
                double overflow = (glass[i][j] - 1.0) / 2.0;
                if(overflow > 0) {
                    glass[i + 1][j] += overflow;
                    glass[i + 1][j + 1] += overflow;
                }
            }
        }
        return min(1.0, glass[query_row][query_glass]);
    }
};

File: 815-champagne-tower/champagne-tower.rs
impl Solution {
    pub fn champagne_tower(poured: i32, query_row: i32, query_glass: i32) -> f64 {
        let mut row: Vec<f64> = vec![0.0; 101];
        let mut next_row: Vec<f64> = vec![0.0; 101];
        row[1] = poured as f64;

        for i in 1..=query_row as usize {
            for j in 1..=i+1 {
                next_row[j] = 0.0_f64.max(row[j - 1] - 1.0) / 2.0 
                            + 0.0_f64.max(row[j] - 1.0) / 2.0;
            }
            std::mem::swap(&mut row, &mut next_row);
        }

        row[query_glass as usize + 1].min(1.0)
    }
}